Add GPCLK support on RP1

P33M opened the pull request:
Prompted by https://forums.raspberrypi.com/viewtopic.php?t=365761&hilit=gpclk this lets users drive gpclks via overlays (and if suitably motivated, drive off-chip sources into peripherals).

I've noticed that clocks without the enable bit set (either by firmware or Linux) report a nonzero frequency - probably a bug, but I don't know where...

P33M committed:
dt-bindings: add additional RP1 PLL output channels

PLL_AUDIO has a ternary divider (a copy of the secondary divider) and
PLL_VIDEO has a primary phased output.

Signed-off-by: Jonathan Bell <jonathan@raspberrypi.com>

P33M committed:
drivers: clk: rp1: add GPCLK source muxes and additional PLL dividers

General-purpose clocks are routed (via a pad) to a large variety of
peripheral aux muxes, and themselves gather a large variety of source
clocks. Entries without a corresponding name string should not be
selected - they bring out internal test/debug clocks which may be
intermittent or very high frequency.

As the GPCLK inputs to peripheral muxes come from a pad, differentiate
the source name from the divider output name. This allows the
possibility of specifying an off-chip clock source to drive the internal
peripheral clock.

Signed-off-by: Jonathan Bell <jonathan@raspberrypi.com>

P33M committed:
DT: rp1: add general-purpose clock source definitions

GPCLKs have two parts - a clock divider and a clock input, routed out to
and in from a GPIO pad respectively. It follows that the clksrc_gpN
inputs can't be used unless the pad is also configured as a GPCLK, so
leave them disabled.

Signed-off-by: Jonathan Bell <jonathan@raspberrypi.com>

pelwell commented on drivers/clk/clk-rp1.c:
```diff
@@ -1745,9 +1818,15 @@ static const struct rp1_clk_desc clk_desc_array[] = {
 					    "pll_audio",
-					    "pll_audio_sec"},
+					    "pll_audio_sec",
+					     "clksrc_gp0",
```
The indentation is off here.

pelwell commented on drivers/clk/clk-rp1.c:
```diff
@@ -1830,9 +1923,25 @@ static const struct rp1_clk_desc clk_desc_array[] = {
+					    "clksrc_gp5",
+					    "pll_sys",
+					    "pll_audio",
+					    "",
```
I think the  "-", used elsewhere, looks nicer - could it work instead?

P33M replied:
The "-" is used as a break/delimiter between the glitchless mux sources and the aux mux sources - the choice of an empty string was deliberate to make the source selection invalid. I'm not sure why the delimiter is necessary - the driver has the information from the num_std_parents and num_aux_parents to work with.

P33M replied:
There's a gotcha that means a delimiter is needed - the mux assignments are not sequential - the second index into the standard sources is always the aux selector (if a glitchless mux is implemented).

pelwell commented on include/dt-bindings/clock/rp1.h:
```diff
@@ -50,3 +50,7 @@
 #define RP1_CLK_MIPI0_DPI		41
 #define RP1_CLK_MIPI1_DPI		42
+
+/* Extra PLL output channels - RP1B0 only*/
```
Missing space.

P33M committed:
drivers: clk: rp1: constrain clock divider outputs to design maximums

Overclocking peripherals is generally a bad thing to do - so reject any
attempt to set a clock output higher than it should be.

Signed-off-by: Jonathan Bell <jonathan@raspberrypi.com>

oOTnTOo commented:
So... How can i use GPCLK0? I use "pinctrl set 4 a0"，but its just "lo" or "hi"。i am try DTS but not work. my system is "Linux raspberrypi 6.6.51+rpt-rpi-2712 #1 SMP PREEMPT Debian 1:6.6.51-1+rpt3 (2024-10-08) aarch64 GNU/Linux", model is Raspberry Pi 5 , bcm 2712。

P33M commented:
Commenting on a closed pull request isn't recommended.

Ask your question in the support forum at: https://forums.raspberrypi.com/viewforum.php?f=29
